### PR TITLE
Add template parameters to Search Applications

### PR DESCRIPTION
**Pull request by carlosdelest**
Original PR https://github.com/carlosdelest/elasticsearch/pull/2
Continued work from https://github.com/kderusso/elasticsearch/pull/4

JSON Schema can be stored as part of the template for a Search Application:

`PUT _application/search_application/<name>`

```
{
    "indices": ["test1"],
    "template": {
        "script": {
            "source": {
                "query": {
                    "term": {
                        "${field_name}": "${field_value}"
                    }
                }
            },
            "params": {
                "field_name": "hello",
                "field_value": "world"
            }
        },
        "dictionary": {
            "properties": {
                "field_name": {
                    "type": "string"
                },
                "field_value": {
                    "type": "string"
                }
            },
            "required": ["field_name"],
            "additionalProperties": false
        }
    }
}
```
dictionary is a JSON Schema used to validate parameters in the _search endpoint in the future.

The dictionary content must be a valid JSON Schema.

Demo script: 

```
# Create indices 
PUT /test1
PUT /test2

# Add docs 
POST /test1/_doc/1
{
  "test": "value",
  "hello": "world",
  "value": 1
}
POST /test2/_doc/2
{
  "test": "another",
  "hello": "world",
  "value": 2
}

# Create search app 
PUT _application/search_application/test-app
{
  "indices": [
    "test1",
    "test2"
  ],
  "template": {
    "script": {
      "source": {
        "query": {
          "term": {
            "{{field_name}}": "{{field_value}}"
          }
        }
      },
      "params": {
        "field_name": "hello",
        "field_value": "world"
      }
    },
    "dictionary": {
            "properties": {
                "field_name": {
                    "type": "string"
                },
                "field_value": {
                    "type": "string"
                }
            },
            "required": ["field_name"],
            "additionalProperties": false
    }
  }
}

GET _application/search_application/test-app

POST /_application/search_application/test-app/_search
{
  "params": {
    "field_name": "hello",
    "field_value": "world"
  }
}
}

POST /_application/search_application/test-app/_search
{
  "params": {
    "field_name": "test",
    "field_value": "another"
  }
}
}
```

### CODE REVIEW

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml [hunk 1]
```diff
@@ -45,6 +45,14 @@ setup:
               params:
                 field_name: field1
                 field_value: value1
+            dictionary:
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml:48]: `dictionary` now allows the full JSON Schema syntax, including `required` parameters and limiting `additionalProperties`.

I think we should change the name from `dictionary` to `param_schema`. It is more descriptive and in line with using a JSON schema for validation.

File: x-pack/plugin/ent-search/build.gradle [hunk 2]
```diff
@@ -23,5 +32,15 @@ dependencies {
   javaRestTestImplementation(project(':modules:lang-mustache'))
 }
 
-addQaCheckDependencies(project)
+tasks.named("dependencyLicenses").configure {
```
Comment (carlosdelest) [inline x-pack/plugin/ent-search/build.gradle:35]: Licenses for external dependencies needed to be added and some tweaking done to avoid repeating licenses for Jackson.

Some ignored classes to be added to audit as well, already done in other ES modules.

File: x-pack/plugin/ent-search/licenses/commons-lang3-LICENSE.txt
Comment (carlosdelest) [inline x-pack/plugin/ent-search/licenses/commons-lang3-LICENSE.txt:1]: Licenses were duplicated where possible from the codebase, or retrieved in the case of json-schema-validator to ensure the license check is passed.

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml
Comment (kderusso) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:1]: Some notes on the file itself: 
- I realize this was here before your PR, but should we take the opportunity to delete the indices created in this test (`test-index1`, `test-index2`, `test-index3`, `test-index4`) in the teardown phase as well? 
- 🤦‍♀️ I thought we'd already taken out the `match: { "foo" }` on line 88, should we clean this up now?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml [hunk 3]
```diff
@@ -109,29 +115,34 @@ teardown:
...
+            script:
+              source: "{ \"query\": { \"query_string\":{ \"query\": \"{{query_string}}\", \"default_field\": \"_all\" } } }"
+              params:
                 query_string: "elastic"
-              },
               lang: "mustache"
-            }
+            dictionary:
+              properties:
+                another_query_string:
```
Comment (kderusso) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:125]: Do we want this request to succeed? Even though it's string encoded, this template is specifying the parameter `query_string` and it's not in the dictionary.

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml [hunk 4]
```diff
@@ -109,29 +115,34 @@ teardown:
...
+  - match:
+      template:
+        script:
+          source: "{ \"query\": { \"query_string\":{ \"query\": \"{{query_string}}\", \"default_field\": \"_all\" } } }"
+          params:
+            query_string: "elastic"
+          lang: "mustache"
+        dictionary:
+          properties:
+            another_query_string:
```
Comment (kderusso) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:141]: Same question as above: Do we want this request to succeed? Even though it's string encoded, this template is specifying the parameter query_string and it's not in the dictionary.

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml [hunk 5]
```diff
@@ -152,25 +163,43 @@ teardown:
...
-          default_field: "*"
-        },
-        lang: "mustache"
-      }
-    }
-  }
+      search_application.get:
+        name: test-updated-search-application
+
+  - match:
```
Comment (kderusso) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:173]: Should we consider including a dictionary with the default template that we provide?

File: x-pack/plugin/ent-search/licenses/slf4j-api-NOTICE.txt
Comment (kderusso) [inline x-pack/plugin/ent-search/licenses/slf4j-api-NOTICE.txt:1]: Is this file required, since it's empty?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml (hunk 1 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml:48]: I like your rename suggestion 👍

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml
Comment (kderusso) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml:1]: Should we add tests to verify that `additionalProperties` works as intended?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/60_behavioral_analytics_list.yml [hunk 6]
```diff
@@ -14,6 +14,11 @@ teardown:
         name: my-test-analytics-collection
         ignore: 404
 
+  - do:
```
Comment (kderusso) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/60_behavioral_analytics_list.yml:17]: Thanks for taking care of this!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java [hunk 7]
```diff
@@ -43,60 +50,75 @@ public class SearchApplicationTemplate implements ToXContentObject, Writeable {
...
             SearchApplication.TEMPLATE_SCRIPT_FIELD
         );
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> {
+            try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
+                return new TemplateParamValidator(builder.copyCurrentStructure(p));
+            }
+        }, DICTIONARY_FIELD);
+    }
+
+    public SearchApplicationTemplate(Script script, @Nullable TemplateParamValidator templateParamValidator) {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java:106]: Why is `templateParamValidator` nullable?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplateService.java [hunk 8]
```diff
@@ -37,9 +37,12 @@ public SearchApplicationTemplateService(ScriptService scriptService, NamedXConte
         this.xContentRegistry = xContentRegistry;
     }
 
-    public SearchSourceBuilder renderQuery(SearchApplication searchApplication, Map<String, Object> templateParams) throws IOException {
+    public SearchSourceBuilder renderQuery(SearchApplication searchApplication, Map<String, Object> templateParams) throws IOException,
+        ValidationException {
```
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplateService.java:41]: Nitpick: 

```suggestion
    public SearchSourceBuilder renderQuery(SearchApplication searchApplication, Map<String, Object> templateParams) 
    throws IOException, ValidationException {
```

**Review by kderusso — COMMENTED** [review]
This looks great @carlosdelest , thank you! I have a few comments and questions mostly surrounding tests. Great work 💯

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml
Comment (carlosdelest) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:1]: Will do!

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml (hunk 3 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:125]: The dictionary does not include the parameters in the template. There are the following reasons for that:
- Default params are included as part of the template creation, so they can be validated by the user that updates the template 
- It's not enough to validate the default params via a dictionary, but they need to be checked with the template as well to ensure it is rendered properly.
- It's easier to implement mandatory params for a template this way, as it's enough to include `required` params in the dictionary. Otherwise, we need to provide a separate mechanism for validating mandatory parameters.

I'm happy to discuss this - for now I think it's a good compromise and allows for the most flexibility in _search params validation. I know @jimczi has thoughts on this as well, so let's discuss - hopefully in a separate PR 😅

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml (hunk 5 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:173]: We could. We can include an array of strings for `default_field` and a string for `query_string`. Changing the template implies changing the dictionary as well, but I think that's to be expected.

Sounds good?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml
Comment (carlosdelest) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/55_search_application_search.yml:1]: I'd say no - we're already checking JSON schema validation works, I think it's enough to test some of its features and not everything it offers.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java (hunk 7 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java:106]: In case there is no `dictionary`, there's no validation. That is what having a `null` `TemplateParamValidator` means.

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml (hunk 3 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:130]: I'm OK with this for the PR, happy to discuss follow up iterations 👍

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml (hunk 5 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:178]: The one issue with an array of Strings is the existing trailing comma issue: https://github.com/elastic/elasticsearch/pull/95200 

As I think about it more - let's punt on changes, and add a dictionary when we update the default template for 8.9.0 to be a bit more sophisticated. WDYT?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java (hunk 7 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java:106]: Got it. I assume we don't want to take a hard line and require a dictionary?

Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java:106]: I can see dictionary being added later in the game, when people is already comfortable with the query they are producing and then want to harden the parameters passed in. I think it stands in the way otherwise.

File: x-pack/plugin/ent-search/licenses/slf4j-api-NOTICE.txt
Comment (carlosdelest) [reply x-pack/plugin/ent-search/licenses/slf4j-api-NOTICE.txt:1]: It is empty in other instances of SLF4J licenses - I think this is a check that both license and notice are included for external deps, even though they might not apply as in this case.

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml (hunk 5 above)
Comment (carlosdelest) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:178]: I'm ok with that - TBH I had not default templates in mind for this change 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java (hunk 7 above)
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationTemplate.java:106]: OK, that makes sense. Thanks for discussing it!

**Review by kderusso — APPROVED** [review]
Nice work @carlosdelest 🙌

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml
Comment (carlosdelest) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:1]: Ah, got it. Removed

Comment (kderusso) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_search_application_put.yml:1]: Thank you!

**Comment by carlosdelest** [comment]
@jimczi , @ioanatia , could I have your review on this? Are we OK with merging this PR and discuss applying the dictionary to  default template params as a separate issue?

**Review by jimczi — APPROVED** [review]
LGTM to unblock
